doorman/jwt_generic.go

```diff
@@ -0,0 +1,189 @@
+package doorman
+
+import (
+	"encoding/json"
+	"fmt"
+	log "github.com/sirupsen/logrus"
+	"io/ioutil"
+	"net/http"
+	"strings"
+	"time"
+
+	"github.com/allegro/bigcache"
+	jose "gopkg.in/square/go-jose.v2"
+	jwt "gopkg.in/square/go-jose.v2/jwt"
+)
+
+// OpenIDConfiguration is the OpenID provider metadata about endpoints etc.
+type OpenIDConfiguration struct {
+	JWKSUri string `json:"jwks_uri"`
+}
+
+// JWKS are the JWT public keys
+type JWKS struct {
+	Keys []jose.JSONWebKey `json:"keys"`
+}
+
+type jwtGenericValidator struct {
+	Issuer             string
+	ClaimExtractor     ClaimExtractor
+	SignatureAlgorithm jose.SignatureAlgorithm
+	cache              *bigcache.BigCache
+}
+
+// newJWTGenericValidator returns a generic JWT validator of this issuer.
+func newJWTGenericValidator(issuer string, extractor ClaimExtractor) *jwtGenericValidator {
+	cache, _ := bigcache.NewBigCache(bigcache.DefaultConfig(1 * time.Hour))
+
+	if extractor == nil {
+		extractor = &defaultClaimExtractor{}
+	}
+
+	return &jwtGenericValidator{
+		Issuer:             issuer,
+		ClaimExtractor:     extractor,
+		SignatureAlgorithm: jose.RS256,
+		cache:              cache,
+	}
+}
+
+func (v *jwtGenericValidator) config() (*OpenIDConfiguration, error) {
+	cacheKey := "config:" + v.Issuer
+	data, err := v.cache.Get(cacheKey)
+
+	// Cache is empty or expired: fetch again.
+	if err != nil {
+		uri := strings.TrimRight(v.Issuer, "/") + "/.well-known/openid-configuration"
+		log.Debugf("Fetch OpenID configuration from %s", uri)
+		data, err = downloadJSON(uri)
+		if err != nil {
+			return nil, err
+		}
+		v.cache.Set(cacheKey, data)
+	}
+
+	// XXX: since cache stores bytes, we parse it again at every usage :( ?
```

Comment:
@mostlygeek I'm running some benchmarks here, and it seems to have a higher cost that I first thought:
```
BenchmarkParseKeys-4   	   20000	     86019 ns/op
PASS
ok  	github.com/mozilla/doorman/doorman	6.148s
```
compared to if I store the parsed data: 
```
BenchmarkParseKeys-4   	1000000000	         2.63 ns/op
PASS
ok  	github.com/mozilla/doorman/doorman	3.208s
```

I come to think that I could manage a minimalist cache code myself (local attribute + TTL) instead of relying on a bytes-only cache solution.

What do you think?

Reply:
86019ns = 0.086ms.  That's still pretty fast, about 11,600 decodes per second on your machine. Pretty sure we won't see that many requests per second. 

So pick cleaner code over weird faster code. If  this ever becomes too slow we can pick [faster json libs](https://github.com/json-iterator/go) to try out.

Reply:
oh you're right! What was I thinking?!? I simply forgot there were µs between ns and ms 🤣